src/main/java/gross/gameoflife/Grid.java

```diff
@@ -0,0 +1,224 @@
+                        }
+                        if (!alreadyUsed) {
+                            // add to cellsToComeAlive if not already there
+                            if (findLiveNeighbors(deadCell[0], deadCell[1]) == 3) {
+                                cellsToComeAlive.add(deadCell);
+                            }
+                            // can be optimized by tracking which cells were already checked
+                        }
+                    }
+                }
+            }
+        }
+        // kill all cells in cellsToDie
+        for (int[] cellToKill : cellsToDie) {
+            setCellDead(cellToKill[0], cellToKill[1]);
+        }
+
+        // revive all cells in cellsToComeAlive
+        for (int[] cellToRevive : cellsToComeAlive) {
+            setCellAlive(cellToRevive[0], cellToRevive[1]);
+        }
+    }
+
+    public int findLiveNeighbors(int row, int column) {
+        int liveNeighbors = 0;
+
+        boolean hasTop = row > 0; // a higher row exists
+        boolean hasBottom = row < board.length - 1; // a lower row exists
+        boolean hasLeft = column > 0; // a left column exists
+        boolean hasRight = column < board[0].length - 1; // a right column exists
+
+        if (hasTop) {
+            // check if cell directly above current cell is alive
+            if (board[row - 1][column] == 1) {
+                liveNeighbors++;
+            }
+            if (hasLeft) {
+                // check if diagonal upper left cell is alive
+                if (board[row - 1][column - 1] == 1) {
+                    liveNeighbors++;
+                }
+            }
+            if (hasRight) {
+                // check if diagonal upper right cell is alive
+                if (board[row - 1][column + 1] == 1) {
+                    liveNeighbors++;
+                }
+            }
+        }
+        if (hasBottom) {
+            // check if cell directly below current cell is alive
+            if (board[row + 1][column] == 1) {
+                liveNeighbors++;
+            }
+            if (hasLeft) {
+                // check if diagonal lower left cell is alive
+                if (board[row + 1][column - 1] == 1) {
+                    liveNeighbors++;
+                }
+            }
+            if (hasRight) {
+                // check if diagonal lower right cell is alive
+                if (board[row + 1][column + 1] == 1) {
+                    liveNeighbors++;
+                }
+            }
+        }
+        if (hasLeft) {
+            // check if adjacent left cell is alive
+            if (board[row][column - 1] == 1) {
+                liveNeighbors++;
+            }
+        }
+        if (hasRight) {
+            // check if adjacent right cell is alive
+            if (board[row][column + 1] == 1) {
+                liveNeighbors++;
+            }
+        }
+
+        return liveNeighbors;
+    }
+
+    public ArrayList<int[]> findDeadNeighbors(int row, int column) {
```

Comment:
What does this method do?

Reply:
The reason I added this was because I figured it would make more sense to identify the dead neighbors of all the live cells and then test those for 3 live neighbors to see if they should be revived. I wrote this method to avoid the extraneous checking of all the dead cells in the grid that do not have any live neighbors to begin with.